File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/evidence/supersession.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Any

ISO_TS = re.compile(
    r"^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?$"
)


def supersession_comparable_source_class(source_type: str) -> str:
    if source_type in {"DOCTOR_STRUCTURED_ENTRY", "DOCTOR_STRUCTURED_PHOTO_OBSERVATION"}:
        return source_type
    return source_comparability_tier(source_type)


def source_comparability_tier(source_type: str) -> str:
    if source_type in {"DOCTOR_STRUCTURED_ENTRY", "DOCTOR_STRUCTURED_PHOTO_OBSERVATION"}:
        return "TIER1_DOCTOR_STRUCTURED"
    if source_type == "DIGITAL_STRUCTURED_REPORT":
        return "TIER2_DIGITAL_STRUCTURED"
    if source_type == "OCR_EXTRACTED_DOCUMENT_IMAGE":
        return "TIER3_OCR_DOCUMENT_IMAGE"
    if source_type == "DOCTOR_FREE_TEXT_NLP_EXTRACTION":
        return "TIER4A_EXTRACTED_CANDIDATE"
    if source_type == "DATASET_TAXONOMY_ALIGNMENT":
        return "TIER4B_SUPPORTING_ONLY"
    return "NOT_COMPARABLE"


def normalize_laterality(laterality: str | None) -> str:
    if laterality is None or str(laterality).strip() == "":
        return ""
    return str(laterality).strip().upper()


def normalized_finding_signature(item: dict[str, Any]) -> str:
    v = item.get("value")
    val = "" if v is None else str(v)
    unit = item.get("unit") or ""
    return f"{val}|{unit}|{item.get('assertion_status')}"


def is_comparable_timestamp(ts: str | None) -> bool:
    if ts is None or str(ts).strip() == "":
        return False
    t = str(ts).strip()
    if not ISO_TS.match(t):
        return False
    try:
        datetime.fromisoformat(t.replace("Z", "+00:00"))
        return True
    except ValueError:
        return False


def supersession_group_key(item: dict[str, Any]) -> str:
    source_class = supersession_comparable_source_class(item.get("source_type", ""))
    test_id = item.get("test_panel_identity") or item.get("finding_identity_key") or item["finding_id"]
    return "|".join(
        [
            item["formula_target_id"],
            item["target_organ_system"],
            item["anatomical_site"],
            normalize_laterality(item.get("laterality")),
            item["target_pathology_id"],
            test_id,
            source_class,
        ]
    )


def apply_supersession(usable_items: list[dict[str, Any]]) -> list[dict[str, Any]]:
    by_group: dict[str, list[dict[str, Any]]] = {}
    for item in usable_items:
        by_group.setdefault(supersession_group_key(item), []).append(item)

    decisions: list[dict[str, Any]] = []
    for group in by_group.values():
        invalid_ts: list[dict[str, Any]] = []
        valid_ts: list[dict[str, Any]] = []
        for item in group:
            if is_comparable_timestamp(item.get("timestamp_or_case_context")):
                valid_ts.append(item)
            else:
                invalid_ts.append(item)

        for item in invalid_ts:
            decisions.append(
                {
                    "finding_id": item["finding_id"],
                    "superseded_by_finding_id": None,
                    "active": True,
                    "reason_codes": ["SUPERSESSION_TIMESTAMP_NOT_COMPARABLE"],
                    "limitation_codes": ["SUPERSESSION_SKIPPED_INVALID_TIMESTAMP"],
                }
            )

        if not valid_ts:
            continue

        sorted_g = sorted(valid_ts, key=lambda x: (x["timestamp_or_case_context"], x["finding_id"]))
        max_ts = sorted_g[-1]["timestamp_or_case_context"]
        at_max = [i for i in sorted_g if i["timestamp_or_case_context"] == max_ts]
        below_max = [i for i in sorted_g if i["timestamp_or_case_context"] != max_ts]
        signatures = {normalized_finding_signature(i) for i in at_max}
        anchor = sorted(at_max, key=lambda x: x["finding_id"])[0]["finding_id"]

        lim = ["SUPERSESSION_SAME_TIMESTAMP_NO_CLINICAL_WINNER"] if len(signatures) > 1 else []
        for item in at_max:
            decisions.append(
                {
                    "finding_id": item["finding_id"],
                    "superseded_by_finding_id": None,
                    "active": True,
                    "reason_codes": [],
                    "limitation_codes": lim,
                }
            )
        for item in below_max:
            decisions.append(
                {
                    "finding_id": item["finding_id"],
                    "superseded_by_finding_id": anchor,
                    "active": False,
                    "reason_codes": ["EVIDENCE_SUPERSEDED"],
                    "limitation_codes": [],
                }
            )

    return sorted(decisions, key=lambda d: d["finding_id"])

File: packages/clinical-contracts/src/rule4/evidence/supersession.ts
import type { Rule4EvidenceItemEnvelope } from './types.js';
import {
  normalizeLateralityForComparability,
  normalizedFindingSignature,
  supersessionComparableSourceClass,
} from './sourceComparabilityTier.js';

const ISO_TS = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?$/;

export function isComparableTimestamp(timestampOrCaseContext: string | null | undefined): boolean {
  if (timestampOrCaseContext == null || String(timestampOrCaseContext).trim() === '') {
    return false;
  }
  const t = String(timestampOrCaseContext).trim();
  if (!ISO_TS.test(t)) {
    return false;
  }
  const ms = Date.parse(t);
  return !Number.isNaN(ms);
}

export function supersessionGroupKey(item: Rule4EvidenceItemEnvelope): string {
  const sourceClass = supersessionComparableSourceClass(item.sourceType);
  const testIdentity = item.testPanelIdentity ?? item.findingIdentityKey ?? item.findingId;
  return [
    item.formulaTargetId,
    item.targetOrganSystem,
    item.anatomicalSite,
    normalizeLateralityForComparability(item.laterality),
    item.targetPathologyId,
    testIdentity,
    sourceClass,
  ].join('|');
}

export type Rule4SupersessionDecision = {
  findingId: string;
  supersededByFindingId: string | null;
  active: boolean;
  reasonCodes: string[];
  limitationCodes: string[];
};

/**
 * Cross-parent supersession when strictly newer timestamp within the same comparability group.
 * Same timestamp: no clinical winner; identical signatures corroborate; conflicts defer to contradiction.
 */
export function applySupersession(
  usableItems: readonly Rule4EvidenceItemEnvelope[],
): Rule4SupersessionDecision[] {
  const byGroup = new Map<string, Rule4EvidenceItemEnvelope[]>();
  for (const item of usableItems) {
    const key = supersessionGroupKey(item);
    const list = byGroup.get(key) ?? [];
    list.push(item);
    byGroup.set(key, list);
  }

  const decisions: Rule4SupersessionDecision[] = [];

  for (const group of byGroup.values()) {
    const invalidTs: Rule4EvidenceItemEnvelope[] = [];
    const validTs: Rule4EvidenceItemEnvelope[] = [];
    for (const item of group) {
      if (isComparableTimestamp(item.timestampOrCaseContext)) {
        validTs.push(item);
      } else {
        invalidTs.push(item);
      }
    }

    for (const item of invalidTs) {
      decisions.push({
        findingId: item.findingId,
        supersededByFindingId: null,
        active: true,
        reasonCodes: ['SUPERSESSION_TIMESTAMP_NOT_COMPARABLE'],
        limitationCodes: ['SUPERSESSION_SKIPPED_INVALID_TIMESTAMP'],
      });
    }

    if (validTs.length === 0) {
      continue;
    }

    const sorted = [...validTs].sort((a, b) => {
      const cmp = a.timestampOrCaseContext.localeCompare(b.timestampOrCaseContext);
      if (cmp !== 0) {
        return cmp;
      }
      return a.findingId.localeCompare(b.findingId);
    });

    const maxTs = sorted[sorted.length - 1]!.timestampOrCaseContext;
    const atMax = sorted.filter((i) => i.timestampOrCaseContext === maxTs);
    const belowMax = sorted.filter((i) => i.timestampOrCaseContext !== maxTs);

    const signaturesAtMax = new Set(atMax.map((i) => normalizedFindingSignature(i)));
    const auditAnchorFindingId = [...atMax].sort((a, b) =>
      a.findingId.localeCompare(b.findingId),
    )[0]!.findingId;

    for (const item of atMax) {
      decisions.push({
        findingId: item.findingId,
        supersededByFindingId: null,
        active: true,
        reasonCodes: [],
        limitationCodes:
          signaturesAtMax.size > 1 ? ['SUPERSESSION_SAME_TIMESTAMP_NO_CLINICAL_WINNER'] : [],
      });
    }

    for (const item of belowMax) {
      decisions.push({
        findingId: item.findingId,
        supersededByFindingId: auditAnchorFindingId,
        active: false,
        reasonCodes: ['EVIDENCE_SUPERSEDED'],
        limitationCodes: [],
      });
    }
  }

  return decisions.sort((a, b) => a.findingId.localeCompare(b.findingId));
}
